bokeh/core/properties.py

```diff
@@ -711,6 +711,38 @@ def __setattr__(self, name, value):
             raise AttributeError("unexpected attribute '%s' to %s, %s attributes are %s" %
                 (name, self.__class__.__name__, text, nice_join(matches)))
 
+    # this is like _value_record_references but expects to find refs
+    # instead of models, and takes a doc to look up the refs in
+    def _json_record_references(doc, v, result, recurse):
+        if v is None: return
+        if isinstance(v, dict) and set(v.keys()) == set(['id', 'type']):
+            if v['id'] not in result:
+                model = doc.get_model_by_id(v['id'])
+                HasProps._value_record_references(model, result, recurse)
+        elif isinstance(v, (list, tuple)):
+            for elem in v:
+                HasProps._json_record_references(doc, elem, result, recurse)
+        elif isinstance(v, dict):
+            for k, elem in v.items():
+                HasProps._json_record_references(doc, elem, result, recurse)
+
+    # add all references from 'v' to 'result', if recurse
+    # is true then descend into refs, if false only
+    # descend into non-refs
+    def _value_record_references(v, result, recurse):
```

Comment:
removed. I am not sure how to re-implement this with. It's also not ever currently getting called? The `value_record` function has
```
if v['id'] not in result:
                result[v['id']] = v
``` 
which should crash if it was called `v` should be a `Model` there